### PR TITLE
Tree arena

### PR DESCRIPTION
**Pull request by KGrewal1**
I've had a try at implementing the unsafe `tree_arena` in a separate lib (but in the same workspace for now) - haven't thought of a good way to make non root access not O(depth), without slowing down insertion or using more memory, though have improved accessing nodes from the root to O(1) from O(depth) and accessing direct children to O(1) from O(children)

### CODE REVIEW

**Comment by KGrewal1** [comment]
I am definitely doing something silly to break the needed preconditions, just need to work out what

**Comment by PoignardAzur** [comment]
> haven't thought of a good way to make non root access not O(depth), without slowing down insertion or using more memory,

The eventual plan will be to use a slotmap.

**Comment by PoignardAzur** [comment]
(Copying from zulip)

Your PR looks good, but we'd probably want more intermediary steps before relying on custom unsafe code; especially since the TreeArena API isn't quite settled yet.

Those steps would be:

- Moving TreeArena to a separate crate without changing the code.
- Adding a test suite.
- Adding a command-line flag so we can switch between the safe implementation and the efficient implementation.

The test suite would help us test the unsafe version with MIRI, which we'd probably add to CI.

**Review by PoignardAzur — CHANGES_REQUESTED** [review]
This is a very high-quality contribution, and will be good to merge once the review points are addressed.

---

Something orthogonal to this review is the question of how the TreeArena API will evolve.

Having an unsafe implementation may slow us down if we want to try API changes and we have to make sure they're implemented in both versions.

I'd like to establish that the safe version will be a first-class citizen and the unsafe version a second-class citizen:

- The safe version may have features / APIs that the unsafe version doesn't yet have.
- If both versions are at feature parity, Masonry can switch on the unsafe version for best performance.
- Otherwise, Masonry uses the safe version.

If that's what we go with, we should documented that pattern in a few places.

File: tree_arena/src/tree_arena_unsafe.rs
Comment (KGrewal1) [reply tree_arena/src/tree_arena_unsafe.rs:64]: It was kept in as I thought that it may be needed to implement tree traversal methods, but was unsure whether they would be useful - noted that https://github.com/linebender/xilem/pull/753 added an iter method over the children of a node

File: tree_arena/src/tree_arena_unsafe.rs
Comment (KGrewal1) [reply tree_arena/src/tree_arena_unsafe.rs:147]: I think unwrap is definitely less preferable (makes more sense to return `None` in the case of the `NodeId` not being in the tree rather than panic). Would you say breaking it down as 

```rust
let node_cell = self.items.get(&id)?;

let TreeNode {
    item,
    children: child_arr,
} = unsafe { node_cell.get().as_ref()? };

let children = ArenaRefChildren {
    parent_arena: self,
    id: Some(id),
    child_arr,
};
Some(ArenaRef {
    parent_id,
    item,
    children,
})
  ```
  is preferable

File: tree_arena/ARCHITECTURE.md
Comment (DJMcNab) [reply tree_arena/ARCHITECTURE.md:33]: I can see an argument for using the GitHub Markdown extension. But for big-O notation, $$O(1)$$ or $$\mathcal{O}(1)$$ is not that much better than O(1)

File: tree_arena/src/tree_arena_unsafe.rs
Comment (KGrewal1) [reply tree_arena/src/tree_arena_unsafe.rs:24]: On further thought, I observed the problem regarding inconsistency with mem::swap when the tests were in the same file (and so accessed the internals) - I think the encapsulation is enough to prevent this as the roots array and the DataMap always belong to the same underlying tree

File: tree_arena/ARCHITECTURE.md
Comment (KGrewal1) [reply tree_arena/ARCHITECTURE.md:33]: On further thinking, given readme is so bare, does it make more sense to flesh this out more, but also move it into Readme rather than having a superfluous file?

File: tree_arena/tests/basic_tests.rs
Comment (KGrewal1) [reply tree_arena/tests/basic_tests.rs:100]: changed this test

File: tree_arena/src/tree_arena_unsafe.rs
Comment (Imberflur) [inline tree_arena/src/tree_arena_unsafe.rs:177]: Hi! I have a minor note on the safety documentation. This is sort of a comment out of nowhere, but I've developed an interest in reviewing unsafe code and I've been following the development in this repo.

Since this retains the full `&mut self` inside the returned `ArenaMutChildren`, I think the safety requirements are more extensive than the case described here. Any use of `parent_arena` in `ArenaMutChildren` needs to be careful to not invalidate the reference produced from `unsafe { item.get().as_mut() }` below. E.g. `DataMap::find` also must only be called on descendants, `parent_arena.items.remove()` must only operate on descendants, `parent_arena.items.insert()` must not overwrite existing nodes, `parent_arena.items` must be used to clear the whole structure, etc.

I was initially thinking the safety documentation could be reworked to note the overall requirement that the `parent_arena` field must not be used to invalidate the returned `item` reference and then include a few examples. However, I think all call sites would have the same safety note pointing to the details `ArenaMutChildren`. So since all the details of `ArenaMutChildren` are private to this module, `find_mut` could be safe, and the `unsafe` block in it could be documented with the fact that all operations that `ArenaMutChildren` allows only access descendants and never invalidate the item reference created here.

Comment (KGrewal1) [reply tree_arena/src/tree_arena_unsafe.rs:177]: Agree with the overall point that the safety requirement should be better put as that you can only access children or remove children of the current node (I'm not sure whether insert() comment would be needed as it is a panic to insert a node with the same name as the key, and not sure clearing is different to removal as there isn't a clear api, and I assume any would be an action on the Arena itself thus checked by the borrow checker as having a reference to an item in the tree prevents any mutable access of the tree itself)

Comment (KGrewal1) [reply tree_arena/src/tree_arena_unsafe.rs:177]: Regarding safety, I think this makes sense, but can also see another argument that it being unsafe is correct in the case it did become public (though assume then the same should be true of the immutable method otherwise could obtain a immutable and mutable reference to the same node)

File: tree_arena/ARCHITECTURE.md
Comment (PoignardAzur) [reply tree_arena/ARCHITECTURE.md]: Yeah. The general workflow I like is "README for things you need to understand the project, ARCHITECTURE for things you need to understand the project's internals that you wouldn't get from the README or the doc root".

File: tree_arena/src/tree_arena_unsafe.rs
Comment (PoignardAzur) [reply tree_arena/src/tree_arena_unsafe.rs:146]: I meant unwrap for `get().as_ref()`, since `UnsafeCell::get()` will never return a null pointer, but either way `?` works. The branches should get elided by the compiler anyway.

> Would you say breaking it down as [...] is preferable?

I'm not a fan of destructuring, but yes, this is much better.

File: tree_arena/src/tree_arena_unsafe.rs
Comment (PoignardAzur) [inline tree_arena/src/tree_arena_unsafe.rs:210]:
```suggestion
    fn get_id_path(&self, id: NodeId, start_id: Option<NodeId>) -> Vec<NodeId> {
```

File: tree_arena/src/tree_arena_unsafe.rs
Comment (PoignardAzur) [inline tree_arena/src/tree_arena_unsafe.rs:340]:
```suggestion
    /// Returns true if there is a child with the given id
```

File: tree_arena/src/tree_arena_unsafe.rs
Comment (PoignardAzur) [inline tree_arena/src/tree_arena_unsafe.rs:403]: Note for the future: we can also avoid the lint by using:

```rust
foobar.unwrap_or_else(|| unreachable!());
```

File: tree_arena/tests/basic_tests.rs
Comment (PoignardAzur) [inline tree_arena/tests/basic_tests.rs:81]: Some of these test failure messages seem redundant.

In general, we tend to assume test assertion failures don't need to be as verbose as API invariant failures, because the former will be read by people familiar with the codebase, whereas the latter will be read by people pulling it as a dependency.

**Review by PoignardAzur — COMMENTED** [review]
LGTM, pending changing the `find` methods and removing the children lists.

You can consider everything else optional before merging.

File: tree_arena/tests/basic_tests.rs
Comment (KGrewal1) [reply tree_arena/tests/basic_tests.rs:81]: I find the advantage of the more verbose test assert is that it is sometimes easier just to grep the assertion failure, to make sure it is the one you think is failing (especially when you make edits and so the line number changes to what is stated), but appreciate this is an odd workflow

Comment (PoignardAzur) [reply tree_arena/tests/basic_tests.rs:81]: > it is sometimes easier just to grep the assertion failure

Oh, that's a good point.

**Comment by PoignardAzur** [comment]
Another blocker before this is merged: I'd like for some version of my comment to show up in the documentation:

> Having an unsafe implementation may slow us down if we want to try API changes and we have to make sure they're implemented in both versions.
> 
> I'd like to establish that the safe version will be a first-class citizen and the unsafe version a second-class citizen:
> 
>     * The safe version may have features / APIs that the unsafe version doesn't yet have.
> 
>     * If both versions are at feature parity, Masonry can switch on the unsafe version for best performance.
> 
>     * Otherwise, Masonry uses the safe version.
> 
> 
> If that's what we go with, we should documented that pattern in a few places.

File: tree_arena/src/tree_arena_unsafe.rs
Comment (Imberflur) [reply tree_arena/src/tree_arena_unsafe.rs]: > I'm not sure whether insert() comment would be needed as it is a panic to insert a node with the same name as the key,

> there isn't a clear api, and I assume any would be an action on the Arena itself thus checked by the borrow checker as having a reference to an item in the tree prevents any mutable access of the tree itself

These are comments about what the safe code in `ArenaMutChildren` is allowed to do with the internal `items` hashmap (which has things such as insert without such checks and clear).

**Comment by KGrewal1** [comment]
Added to the readme and the doc comment at the root of the crate

**Review by PoignardAzur — APPROVED** [review]
LGTM.

I'd maybe like a comment noting that `TreeNode::children` needs to be removed but otherwise we're good.

**Comment by KGrewal1** [comment]
I'm not sure `TreeNode::children` can be removed yet as it is being used on node removal (as we remove all successor nodes) - I think the alternative would be iterating over the whole hashmap to find which children have the current node as their parent but that would definitely be inefficient?

File: tree_arena/src/tree_arena_unsafe.rs
Comment (DJMcNab) [inline tree_arena/src/tree_arena_unsafe.rs:20]: Would we be able to have something like an `UnsafeMutex` type, ideally which is implemented safely by default but with an unsafe passthrough version, to check some that invariants are being met.
That is, have this same implementation, but in a checked manner?

**Review by DJMcNab — COMMENTED** [review]
I don't want to approve the unsafe code, as I don't fully understand it.

However, given that we use the entirely safe version by default, it seems low-risk enough to land this.

File: tree_arena/src/tree_arena_unsafe.rs
Comment (KGrewal1) [reply tree_arena/src/tree_arena_unsafe.rs:20]: This does make some sense but will have the issue that jumping to any child node from the root would require iterating through all the parents along the way (and would need to determine how removal interacts with this as would need to reindex possibly all but the node changed) - think this could also be implemented in the safe version as then parents own their children and so unsafe is not needed. 

Advantage of having the nodes stored in the hashmap is that when it is determined a node is a descendant, can use the map to jump straight to that node. 

Was also thinking whether separate parent and node hashmaps are actually needed or if it would work storing <Option<NodeId>, Box<Node>>, so only one lookup is used

File: tree_arena/src/lib.rs
Comment (KGrewal1) [reply tree_arena/src/lib.rs:4]: For clarity is that the content after the licence in the other lib.rs eg lines 4-22 in https://github.com/linebender/xilem/blob/main/xilem_core/src/lib.rs ?

Comment (DJMcNab) [reply tree_arena/src/lib.rs:4]: Ah, right. This is confusing.

No, this is the setup used specifically in the Masonry crate. The version in Xilem Core is an old version of that

File: tree_arena/src/tree_arena_unsafe.rs
Comment (DJMcNab) [reply tree_arena/src/tree_arena_unsafe.rs:20]: > iterating through all the parents along the way

Why? Wouldn't you be able to jump directly to the specified node's parent?

Comment (KGrewal1) [reply tree_arena/src/tree_arena_unsafe.rs:20]: Sorry yes, misinterpreted the suggestion - this should work I think (having the hashmap associate parents with a collection of their children) - to avoid the indexing issue on node removal, guess a HashSet or nested HashMaps could be used, tho I agree that a vec would be ideal, just need to ensure that indexing stays consistent!

Comment (KGrewal1) [reply tree_arena/src/tree_arena_unsafe.rs:20]: Yes but would require quite a few changes as the TreeNode contains the contents themselves and the list of children, but the API returns these in separate structs - the` item` and the `ArenaChildren` (mut or ref) so I think both would  need to be wrapped in some form of mutex for run time checking separately ?

File: tree_arena/Cargo.toml
Comment (KGrewal1) [reply tree_arena/Cargo.toml:34]: I think I have now changed this? Using the same as in the other crates in the workspace but not yell pulling in examples as have not yet added a folder for them

File: tree_arena/Cargo.toml
Comment (DJMcNab) [reply tree_arena/Cargo.toml:31]: Yes, that's it.

Comment (KGrewal1) [reply tree_arena/Cargo.toml:31]: Lines have been moved around but added to the features section!

**Comment by DJMcNab** [comment]
I've done some spot checks of a few examples, and this all seems to work. As discussed on Zulip, we'll land this now, because the unsafe code is off by default.

**Pull request merged**
